### ISSUE TITLE
DictionaryFilters can re-use StreamBytesInput when compressed

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#2001]
When using an And or Or DictionaryFilter, we re-use the BytesInput across reads. This is problematic when compressed because compressed BytesInputs get converted over to StreamBytesInputs which can only be used once.

**Reporter**: [Patrick Woody](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pwoody) / @pwoody
**Assignee**: [Ryan Blue](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rdblue) / @rdblue
#### PRs and other links:
- [PR #376](https://github.com/apache/parquet-mr/pull/376)

<sub>**Note**: *This issue was originally created as [PARQUET-743](https://issues.apache.org/jira/browse/PARQUET-743). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Patrick Woody](https://issues.apache.org/jira/browse/PARQUET-743?#comment-15555870) / @pwoody:
Copied from mailing list
```Java
Hey all,

Running a parquet-mr build off of master and I'm seeing some interesting behavior when using a DictionaryFilter to prune row groups. Basically, if I have an And or Or filter the DictionaryPage object gets re-used. This seems to be a problem for StreamBytesInput because the stream gets exhausted after the first toByteArray call. My current workaround is to synchronize and just re-use the byte array after the first read, but I'd be curious as to what people think the best approach to solving this is and if we should be reusing the BytesInput at all.

Best,
Patrick 


---

Looking a bit more - it looks like this is because decompression converts to a StreamBytesInput automatically. The current tests run with the uncompressed codec, so it doesn't hit this issue. I've put up a commit here that demonstrates the issue and my current workaround: https://github.com/palantir/parquet-mr/pull/10/commits/70cc00cba5c294d4c860bd4cd2c48c2d083a5809.

Thanks,
```

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-743?#comment-15570431) / @rdblue:
Merged PR #376. Thanks for reviewing, @julienledem!
